pkg/kubectl/apis/manifest/v1alpha1/types.go

```diff
@@ -0,0 +1,123 @@
+/*
+Copyright 2017 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package v1alpha1
+
+import (
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+// Package could be either a base or a overlay manifest.
+type Package struct {
+	metav1.TypeMeta `json:",inline" yaml:",inline"`
+	// Standard object's metadata.
+	// More info: https://git.k8s.io/community/contributors/devel/api-conventions.md#metadata
+	metav1.ObjectMeta `json:"metadata,omitempty" yaml:"metadata,omitempty"`
+
+	// Labels to add to all objects and selectors.
+	// These labels would also be used to form the selector for apply --prune
+	// Named differently than “labels” to avoid confusion with metadata for this object
+	ObjectLabels map[string]string `json:"objectLabels,omitempty" yaml:"objectLabels,omitempty"`
+
+	// Annotations to add to all objects.
+	ObjectAnnotations map[string]string `json:"objectAnnotations,omitempty" yaml:"objectAnnotations,omitempty"`
+
+	// Partial name that will prefix the name of the base resources.
+	NamePrefix string `json:"namePrefix,omitempty" yaml:"namePrefix,omitempty"`
+
+	// Description of the application package.
+	Description string `json:"description,omitempty" yaml:"description,omitempty"`
+
+	// Search keywords for this application package.
+	Keywords []string `json:"keywords,omitempty" yaml:"keywords,omitempty"`
+
+	// Version of the application package.
+	AppVersion string `json:"appVersion,omitempty" yaml:"appVersion,omitempty"`
+
+	// Homepage of the application package.
+	Home string `json:"home,omitempty" yaml:"home,omitempty"`
+
+	// Sources specify the URLs, e.g. https://github.com/foo/bar.git, file://host/path, etc.
+	// hosting the resource files specified in Base and Overlays.
+	Sources []string `json:"sources,omitempty" yaml:"sources,omitempty"`
+
+	// An pointer to the icon.
+	Icon string `json:"icon,omitempty" yaml:"icon,omitempty"`
+
+	// People that maintain this package.
+	Maintainers []Maintainer `json:"maintainers,omitempty" yaml:"maintainers,omitempty"`
+
+	// A Base entry specifies the file path, relative to Source, of an API resource definition.
```

Comment:
These strings should be in whatever form kubectl's --filename parameter accepts. So at least files and directories. I recommend relative paths within the package, so that they'll work no matter how the bundle of files is delivered: fork from github, tarball from registry, curled from object storage, etc. URLs: Have to think more about that. As mentioned in #24649, I'd also like to support globs here.

Explicit individual files would be useful in the case that the package were retrieved directly via HTTP (though we shouldn't require that) or other yaml/json files were present in directories containing the package files.

Reply:
How do you plan to process send those files to the API ? as separate API call for each resources? 
I assume that expansion, generation, transformation, patches are done client-side, in that case what about creating a new field 'resources' to capture that ?  
Server-side would apply/create from this field (like `kind: list`)

Reply:
> URLs: Have to think more about that.

Yes. I have concern of it as well. The underlying content of a URL may change and the user may not know. IMO it's not safe for fork/rebase workflow.

> How do you plan to process send those files to the API ? as separate API call for each resources?
I assume that expansion, generation, transformation, patches are done client-side, in that case what about creating a new field 'resources' to capture that ?
Server-side would apply/create from this field (like kind: list)

Current plan is to have a `expand` command for prototyping. It will yield the api resources after applying the overlay on top of base. I imagine that it can be output to stdout or a file the resources are separated by `---` (I think it is easier than using a list).

I will push more change to https://github.com/kubernetes/kubernetes/pull/52682 soon.